File: java/src/main/java/io/nayuki/qrcodegen/BitBuffer.java
/* 
 * QR Code generator library (Java)
 * 
 * Copyright (c) Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/qr-code-generator-library
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

package io.nayuki.qrcodegen;

import java.util.BitSet;
import java.util.Objects;


/**
 * An appendable sequence of bits (0s and 1s). Mainly used by {@link QrSegment}.
 */
public final class BitBuffer implements Cloneable {
	
	/*---- Fields ----*/
	
	private BitSet data;
	
	private int bitLength;  // Non-negative
	
	
	
	/*---- Constructor ----*/
	
	/**
	 * Constructs an empty bit buffer (length 0).
	 */
	public BitBuffer() {
		data = new BitSet();
		bitLength = 0;
	}
	
	
	
	/*---- Methods ----*/
	
	/**
	 * Returns the length of this sequence, which is a non-negative value.
	 * @return the length of this sequence
	 */
	public int bitLength() {
		assert bitLength >= 0;
		return bitLength;
	}
	
	
	/**
	 * Returns the bit at the specified index, yielding 0 or 1.
	 * @param index the index to get the bit at
	 * @return the bit at the specified index
	 * @throws IndexOutOfBoundsException if index &lt; 0 or index &#x2265; bitLength
	 */
	public int getBit(int index) {
		if (index < 0 || index >= bitLength)
			throw new IndexOutOfBoundsException();
		return data.get(index) ? 1 : 0;
	}
	
	
	/**
	 * Appends the specified number of low-order bits of the specified value to this
	 * buffer. Requires 0 &#x2264; len &#x2264; 31 and 0 &#x2264; val &lt; 2<sup>len</sup>.
	 * @param val the value to append
	 * @param len the number of low-order bits in the value to take
	 * @throws IllegalArgumentException if the value or number of bits is out of range
	 * @throws IllegalStateException if appending the data
	 * would make bitLength exceed Integer.MAX_VALUE
	 */
	public void appendBits(int val, int len) {
		if (len < 0 || len > 31 || val >>> len != 0)
			throw new IllegalArgumentException("Value out of range");
		if (Integer.MAX_VALUE - bitLength < len)
			throw new IllegalStateException("Maximum length reached");
		for (int i = len - 1; i >= 0; i--, bitLength++)  // Append bit by bit
			data.set(bitLength, QrCode.getBit(val, i));
	}
	
	
	/**
	 * Appends the content of the specified bit buffer to this buffer.
	 * @param bb the bit buffer whose data to append (not {@code null})
	 * @throws NullPointerException if the bit buffer is {@code null}
	 * @throws IllegalStateException if appending the data
	 * would make bitLength exceed Integer.MAX_VALUE
	 */
	public void appendData(BitBuffer bb) {
		Objects.requireNonNull(bb);
		if (Integer.MAX_VALUE - bitLength < bb.bitLength)
			throw new IllegalStateException("Maximum length reached");
		for (int i = 0; i < bb.bitLength; i++, bitLength++)  // Append bit by bit
			data.set(bitLength, bb.data.get(i));
	}
	
	
	/**
	 * Returns a new copy of this buffer.
	 * @return a new copy of this buffer (not {@code null})
	 */
	public BitBuffer clone() {
		try {
			BitBuffer result = (BitBuffer)super.clone();
			result.data = (BitSet)result.data.clone();
			return result;
		} catch (CloneNotSupportedException e) {
			throw new AssertionError(e);
		}
	}
	
}

File: csharp/QrCodeGenerator/BitBuffer.cs
using System;
using System.Collections;
using System.Diagnostics;
using Util;

namespace QrCodeGenerator
{
    /// <summary>
    /// An appendable sequence of bits (0s and 1s). Mainly used by <see cref="QrSegment"/>.
    /// </summary>
    public class BitBuffer
    {
        private BitSet Data;
        private int BitLength;

        /// <summary>
        ///  Constructs an empty bit buffer (length 0).
        /// </summary>
        public BitBuffer() 
        {
            Data = new BitSet();
            BitLength = 0;
        }

        /// <summary>
        /// Returns the length of this sequence, which is a non-negative value.
        /// </summary>
        /// <returns>The length of this sequence</returns>
        public int GetBitLength()
        {
            return BitLength;
        }

        /// <summary>
        /// Returns the bit at the specified index, yielding 0 or 1.
        /// </summary>
        /// <param name="index">The index to get the bit at</param>
        /// <returns>The bit at the specified index</returns>
        /// <exception cref="IndexOutOfRangeException">If index &lt; 0 or index &#x2265; bitLength</exception>
        public int GetBit(int index)
        {
            if (index < 0 || index >= BitLength) 
                throw new IndexOutOfRangeException();
            return Data.Get(index) ? 1 : 0;
        }

        /// <summary>
        /// Appends the specified number of low-order bits of the specified value to this
        /// buffer. Requires 0 &#x2264; len &#x2264; 31 and 0 &#x2264; val &lt; 2<sup>len</sup>.
        /// </summary>
        /// <param name="value">The value to append</param>
        /// <param name="length">The number of low-order bits in the value to take</param>
        /// <exception cref="ArgumentException">If the value or number of bits is out of range</exception>
        /// <exception cref="InvalidOperationException">If appending the data</exception>
        public void AppendBits(int value, int length)
        {
            if (length < 0 || length > 31 || value >> length != 0) 
                throw new ArgumentException("Value out of range");
            if (int.MaxValue - BitLength < length) 
                throw new InvalidOperationException("Maximum length reached");

            for (int i = length - 1; i >= 0; i--, BitLength++)
                Data.Set(BitLength, QrCode.GetBit(value, i));
        }

        /// <summary>
        /// Appends the content of the specified bit buffer to this buffer.
        /// </summary>
        /// <param name="bitBuffer">The bit buffer whose data to append (not <c>null</c>)</param>
        /// <exception cref="ArgumentNullException">If the bit buffer is <c>null</c></exception>
        /// <exception cref="InvalidOperationException">If appending the data would make bitLength exceed Integer.MaxValue</exception>
        public void AppendData(BitBuffer bitBuffer)
        {
            if (bitBuffer == null) 
                throw new ArgumentNullException("BitBuffer is null!"); 
            if (int.MaxValue - BitLength < bitBuffer.GetBitLength()) 
                throw new Exception("Maximum length reached");
            
            for (int index = 0; index < bitBuffer.GetBitLength(); index++, BitLength++)
                Data.Set(BitLength, bitBuffer.Data.Get(index));
        }

        /// <summary>
        /// Returns a new copy of this buffer.
        /// </summary>
        /// <returns>A new copy of this buffer (not <c>null</c>)</returns>
        public BitBuffer Clone()
        {
            var result = new BitBuffer().Clone();
            result.Data = (BitSet)result.Data.Clone();
            return result;
        }
    }
}
